store/src/main/java/com/jinyframework/keva/store/IndexStoreImpl.java

```diff
@@ -298,13 +298,18 @@ public void put(String key, Long value) {
     }
 
     protected int getHashBucketOffset(String key) {
-        int offset = -1;
+        int offset = -1; // Set default value to -1 means key not exist
 
         try {
             offset = getHashBucket(key.hashCode());
             indexBuffer.position(offset);
             byte occupied = indexBuffer.get();
-            if (occupied > 0) {
+            // "occupied" is a active flag of every record, type: byte (0 | 1)
```

Comment:
"occupied" is a active flag -> "occupied" is `an` active flag

besides, use 1 byte for the whole flag is quite expensive, use the most significant bit of any other metadata instead

Reply:
@duynguyen-ori75 I'm using this structure for storing records:

![image](https://user-images.githubusercontent.com/13906546/118568863-ea64ed80-b7a2-11eb-8c89-38aa8e4184e1.png)

> Use the most significant bit of any other metadata instead
Please explain more

Reply:
For the flag variables (true/false only), I cannot use `Boolean`, because of the limitation of `java.nio.ByteBuffer` - which allows write and read all primitive types except boolean.

This is an off-heap store. I'm using `ByteBuffer.allocateDirect` to ask directly to the OS for memory (which then be retrieved as a sequence of bytes called byteArray).

Reply:
This is just a "workaround" to deal with boolean values in the ByteBuffer array, so it's not optimal. 😢

Reply:
Actually, both Active + Type can be embedded in the Length metadata, but it is quite tricky to implement this.

Instead, we can use the most significant bit of the type metadata to store the active status and use the rest (7 bytes) to store the real type information

https://www.geeksforgeeks.org/find-significant-set-bit-number/

Besides, why did you select the above structure to store data?

Reply:
Resolved by #9